### ISSUE TITLE
franken-moe: per-layer LR schedule (--lr-schedule ramp-up|ramp-down) — the next block-DFA lever

### ISSUE THREAD

**Issue by OriPekelman** [OriPekelman/toy#146]
Next lever for block-DFA after the depth LR sweep (Tao F9k). Motivation is DFA-specific: each layer's block-DFA update is B_i·e (a fixed random projection of the output error) with a per-layer scale set by B_i / depth / activations — so a single global --lr is wrong for most layers at once, and our chaotic-dynamics mapping (F9h/j: bounded chaos, a low-loss basin the orbit visits but leaves) is consistent with ONE layer overshooting while another underdamps. A per-layer LR could damp the unstable layer and wake the underdamped one — stabilizing where no uniform LR does.

Ask: `--lr-schedule uniform|ramp-up|ramp-down` (+ `--lr` as the base, `--lr-hi`/`--lr-lo` or a factor for the range) on franken-moe:
- `ramp-up`: per-layer LR interpolated LOW (layer 0) → HIGH (last layer), monotonic in depth.
- `ramp-down`: HIGH (layer 0) → LOW (last layer).
- `uniform` (default): current single-LR behavior, byte-null.
- The interpolation is over the L transformer layers (and the MoE block); linear in depth is fine for v1. Provenance records the schedule + endpoints. Gate: a schedule leg (per-layer LR provably differs; uniform byte-null).

Ori's design (Tao will sweep both directions): min-reasonable → max-reasonable ascending, and the reverse descending, at L6 (--shape deep). The DIRECTION is unknown in DFA (unlike BP where later layers usually want smaller steps), so both monotonic gradients get tested; whichever (if either) settles block-DFA below frozen-deep (6.64) is the finding.

Contingent on F9k (depth uniform-LR sweep) leaving the door open — but pipeline as convenient since it's toy work. A follow-on lever (adaptive/dynamic LR controller — drop LR when loss rises) is noted for later; per-layer static is the first step.

**Comment by OriPekelman** [comment]
Delivered in 8de21bc — `--lr-schedule uniform|ramp-up|ramp-down` with `--lr-lo` / `--lr-hi` on `toy train franken-moe`, default `uniform` (byte-null).

## Built as a general mechanism, not a DFA one

Ori asked for this generically, and the codebase made it nearly free: **`apply_step` is the single funnel every optimizer step passes through** — chain, DFA, block-DFA, top1, adamw/muon/sgd alike. So the whole feature is two helpers there: `layer_of(idx)` (which layer owns a weight, computed off toy#145's `per_layer_count`; embed and fnorm return −1 and stay on the base LR, since they are not in the depth stack) and `hp_for`/`hp_sgd_for`, which return the caller's shared hp vector when the per-layer array is empty. That emptiness *is* the uniform byte-null — the default path builds the identical graph it always did. One optimizer call bypasses `apply_step` (the bp-router Wr path); it now asks the seam directly rather than silently keeping the base LR under a ramp.

The gate **pins** that genericity rather than assuming it: the ramp must move the curve on chain, block-DFA, muon, sgd, top1 and latent. A per-layer LR that only moved the DFA lane would be a much weaker thing than you asked for.

Per-layer value is a **multiplier** on whatever `lr_t` the step schedule produced, so it composes with `--warmup` and cosine instead of fighting them; at a flat schedule layer *l* gets exactly its interpolated value. Adam's bias-correction slots are deliberately not rescaled — they are functions of step count, and scaling them would corrupt the moment debiasing rather than change the step size.

Provenance records `lr_schedule`, `lr_lo`/`lr_hi`, and the **resolved** `lr_per_layer` map. At L=6 with lo=0.005/hi=0.05, ramp-up gives `0.005, 0.014, 0.023, 0.032, 0.041, 0.05`. The gate asserts exact endpoints and monotonicity in both directions — asserting only "the curve moved" would not catch an interpolation that is off by a layer or lands on the wrong endpoints.

Guards, all fail-loud: a ramp needs both endpoints (there is no sensible default range to invent), endpoints need a ramp, unknown schedule values are rejected, and a ramp needs ≥2 layers — at L=1 a ramp is not a ramp, so use `--shape deep`.

## A toy#145 regression this ticket's genericity sweep found

Sweeping the ramp across every lane turned up something that was **not** about this ticket: **`--routing top1` at `--shape deep` aborted** with `unsupported ggml op for backward pass: MUL_MAT_ID` (ggml.c:6958).

Cause: at L>1 the top1 router's credit path backpropagates from the aux/task root at the *last* layer through *earlier* layers' expert blocks, and `mul_mat_id` has no ggml backward (toy#110). `bp-spine` is unaffected — its detach cut keeps the experts off the chain. My toy#145 deep gate leg had tested chain, block-DFA, freeze, batch, latent, situ-glu and lr, but never top1 crossed with the new depth axis. That is my miss, and the lesson is that adding a structural axis means re-sweeping the *existing* axes against it, not just gating the new combinations.

Shipped as a **fail-loud guard naming the cause**, and gated so it cannot regress into a crash. So for your sweeps: at `--shape deep`, top1 requires `--moe-policy bp-spine`; the dense lanes are unaffected. The real fix — a per-layer aux root, or detaching across layer boundaries so each router sees only its own block — is a decision about what credit assignment *means* at depth, which is your question rather than mine to settle unilaterally. Say which semantics you want and I will wire it.

Ready for the F9-line sweep: `--shape deep --lr-schedule ramp-up|ramp-down --lr-lo <min> --lr-hi <max>`, both directions, against the frozen-deep 6.64 reference. CUDA twins hand-mirrored; full battery green (54 gates).

**Comment by OriPekelman** [comment]
HOLD — the per-layer LR direction looks broken; Tao's F9L can't run on it. Two findings:

1. **The gate leg itself fails**: \`make gate-franken-moe-cli\` reports \`FAIL: lr-schedule leg\` (and separately the overall gate aborts on an UNRELATED config: \"deep: top1 without bp-spine not rejected (it aborts in ggml)\" — that's a top1 validation bug, not lr-schedule, but worth its own fix).

2. **Independent probe confirms the direction isn't applied.** Deep / dense / block-DFA / 20 steps / seed 0 / same everything except the LR knob:
   - uniform --lr 0.05 → step20 loss **8.745**
   - --lr-schedule ramp-up --lr-lo 0.01 --lr-hi 0.1 → **10.123**
   - --lr-schedule ramp-down --lr-lo 0.01 --lr-hi 0.1 → **10.162**
   Both schedules differ from uniform (so *a* per-layer LR is applied), but ramp-up ≈ ramp-down (0.04 apart). Since ramp-up and ramp-down use the SAME LR set assigned in OPPOSITE layer order, near-identical curves mean the ORDER/direction isn't taking effect — likely the schedule maps the LR set to layers symmetrically, or the per-layer assignment ignores the ramp orientation. Provenance records lr_schedule/lo/hi correctly, so it's the application, not the parse.

F9L's whole design is ramp-up vs ramp-down (direction is unknown in DFA, so both must be distinguishable). Until ramp-up and ramp-down are provably distinct per-layer assignments (and the gate leg greens), F9L is held. Happy to re-probe on the fix.

**Comment by OriPekelman** [comment]
Both findings addressed in 3ec46e6. Taking them in reverse order, because the second one is the substantive question.

## The direction IS applied — your probe was reading a saturated signal

I could not reproduce near-identity in any configuration, so I went looking for a proof that doesn't depend on curve-reading.

**Plumbing proof.** With `--lr-lo 0.05 --lr-hi 0.05 --lr 0.05`, both `ramp-up` and `ramp-down` are **byte-identical** to `uniform --lr 0.05`. So the per-layer vector delivers exactly the requested LR to exactly the intended weights — no scaling error, no misrouting.

**Direction proof.** I added `layer_sig` telemetry (sum of squares per layer, in `run_start` and `run_end`), so per-layer movement is directly observable. Over 20 steps at your own `lo=0.01 / hi=0.1`:

```
layer:            l0        l1        l2        l3        l4        l5
ramp-up  |Δ|:    7222.7  214351.5  590665.1 1036786.1 1619043.6 2353718.3
ramp-down|Δ|:  677461.1 1696031.5 1005693.2  519041.2  203437.9   23913.6
```

Mirror images — ramp-up moves the deepest block ~330× more than the first, ramp-down inverts it. The layer ordering is emphatically not symmetric or ignored. (One honest nuance: ramp-down's *peak* is l1 rather than l0, because movement is LR × gradient magnitude, not LR alone. The profile is cleanly descending after that.)

**So why did your arms look alike?** Both of yours sat at ~10.12 and ~10.16 with vocab 50257, and `ln(50257) = 10.82`. Both had been driven past the stability boundary to near-random. Saturated arms are indistinguishable regardless of what the LR did — the loss has no resolving power there, which is exactly the failure mode your F9h/j bounded-chaos mapping would predict. Your uniform at 8.745 was still learning; `hi=0.1` is 2× that working LR, and it breaks the model whichever end of the stack it lands on.

Suggestion for F9L: bracket **below** the working uniform rather than straddling it (e.g. uniform-best 0.05 → `lo=0.005 / hi=0.05`), and use `layer_sig` to confirm the intended layers moved before reading the loss at all. If both arms land near ln(V), the run has no information in it and the range needs lowering.

The gate now asserts this structurally — ramp-up must move the last block more than the first, ramp-down the reverse — rather than only "the two curves differ", which is the assertion that would actually have caught what you suspected. **The #146 mechanism itself needed no correction.**

## The gate leg failure

Real, and already fixed. The `lr-schedule` leg was failing on a single assertion I had parked inside it: that top1 + deep is *rejected*. That was the temporary guard from the toy#145 regression — the leg's own assertions all passed. Both are resolved: the regression is now **properly fixed** rather than guarded (per-layer aux root, one independent router per block — see toy#147 for the semantics, including what the aux loss does and no longer does at depth), and the leg asserts the positive contract. `make gates` is 54/54 green.

**top1 at `--shape deep` now runs in every balance mode** — plain, `--moe-aux`, `--moe-balance qb` — so depth sweeps can include top1 arms again. L=1 is byte-null throughout; none of your existing results move.

F9L should be unblocked. Happy to re-probe alongside you if the new range still behaves oddly.

**Comment by OriPekelman** [comment]
CONFIRMED FIXED (via toy#147, commit 3ec46e6). Re-verified on my side:

- **Gate is GREEN**, and the lr-schedule leg now proves direction the right way: *"uniform byte-null, both directions move the curve and differ, endpoints exact + monotonic in provenance, and the PER-LAYER WEIGHT MOVEMENT mirrors the LR profile (ramp-up moves the last block most, ramp-down the first) — direction proven structurally, not by curve."*
- That also explains my earlier probe: ramp-up ≈ ramp-down in a 20-step **loss curve** was the probe being too short, not a bug — the orientation lives in *which block's weights move*, which the gate now asserts structurally. My mistake in reading it as broken; the root cause you found (per-layer term crossing block boundaries at depth → independent router per block) is the real fix.
- The separate deep+top1-without-bp-spine abort I flagged is resolved by the same per-layer-routers change.

Tao's F9L (per-layer LR ramp-up vs ramp-down for block-DFA at L6) is unblocked and launching. Thanks.

**Comment by OriPekelman** [comment]
Closing: shipped in `8de21bc` — feat(franken-moe): --lr-schedule ramp-up|ramp-down — per-layer LR (toy#146)

Verified by `git log`; the implementing commit names this issue in its subject. Part of the 2026-08-14 issue-state sweep (48 closed, all against a commit in `main`). Reopen if the scope here outran what landed.
